tools/topology/topology2/doc/topology2-filter.py

```diff
@@ -0,0 +1,456 @@
+	attributes = dict()
+	attrib_doxy = dict()
+	bsum = 1
+	doxy=""
+	includes = ""
+	defaults = io.StringIO()
+	objs = dict()
+	for line in input:
+		if parse_include_str(line):
+			includes = includes + parse_include_str(line)
+			continue
+		if parse_include_by_key(line, input, sys.stdout):
+			continue
+		if re.search("^\s*DefineAttribute\.", line):
+			parse_class_attribute(attributes, doxy, input, line)
+			doxy = ""
+		elif re.search("^\s*DefineArgument\.", line):
+			for line in input: # Just skip, this is only used in bytes.conf
+				if cbracket_count(line) < 0:
+					break
+		elif re.search("^\s*attributes\s+\{", line):
+			parse_attributes_block(input, attributes, attrib_doxy)
+		elif re.search("^\s*Object\.", line):
+			objs = parse_object(input, line, file = defaults, objects = objs,
+					    tabs = "\t", ending = ",")
+		else:
+			(doxy, line) = doxy_check_add(doxy, line)
+			tok = line.split()
+			if len(tok) == 2 and tok[0].isidentifier():
+				print("\t.%s =\t%s," % (trim_c_id(tok[0]), trim_value(tok[1])),
+				      file = defaults)
+			bsum = bsum + cbracket_count(line)
+			if bsum < 1:
+				break
+	attribute_block_print(className, attributes, attrib_doxy)
+	print("struct %s {" % className)
+	print(includes)
+	print_attributes(attributes)
+	for obj in objs:
+		if objs[obj] > 1:
+			for i in range(objs[obj]):
+				print("\tstruct %s %s%d;" % (obj, obj, i))
+		else:
+			print("\tstruct %s %s;" % (obj, obj))
+	print("};\n")
+	print("//! \\var struct %s %s_defaults" % (className, className))
+	print("//! \\brief %s class default values" % className)
+	print("struct %s %s_defaults = {\n" % (className, className))
+	print(defaults.getvalue())
+	print("};\n");
+	defaults.close()
+
+def parse_members(input, cbsum, tabs, file):
+	doxy = ""
+	for line in input:
+		if parse_include_by_key(line, input, file):
+			continue
+		cbsum = cbsum + cbracket_count(line)
+		(doxy, line) = doxy_check_add(doxy, line)
+		# Assume ending } to be alone on its own line
+		if cbsum < 1:
+			return
+		if re.search("^\s*Object\.", line):
+			print("%s: object-line: %s" % (fname(), line), file = debug)
+			obj = line.split(".")
+			if cbsum == 2:
+				# Assume Object.Base.name.1 {
+				doxy = print_doxy(doxy, file = file)
+				print("%s.%s = {" % (tabs, obj[2]), file = file)
+				parse_members(input, 1, tabs + "\t", file = file)
+				cbsum = 1
+				print("%s}," % tabs, file = file)
+			elif cbsum == 1:
+				# Assume  Object.Base.name.1 {}
+				doxy = print_doxy(doxy, file = file)
+				print("%s.%s = {}," % (tabs, obj[2]), file = file)
+		else:
+			tok = line.split(maxsplit = 1)
+			if len(tok) >=2:
+				name = tok[0]
+				val = trim_value(tok[1])
+				doxy = print_doxy(doxy, file = file)
+				print("%s.%s = %s," % (tabs, name, val), file = file)
+
+def object_instance_prefix(name, ending):
+	if ending == ";":
+		return "struct " + name + " "
+	else:
+		return "."
+
+def parse_object(input, fline, file = sys.stdout, objects = dict(), tabs = "", ending = ";"):
```

Comment:
some big functions here, can we have inline comments for each code block and a comment description of each function. i.e. what does function do.